WebDriverAgentLib/Commands/FBSessionCommands.m

```diff
@@ -150,14 +152,46 @@ + (NSArray *)routes
         || [capabilities[FB_CAP_SHOULD_WAIT_FOR_QUIESCENCE] boolValue];
       app.launchArguments = (NSArray<NSString *> *)capabilities[FB_CAP_ARGUMENTS] ?: @[];
       app.launchEnvironment = (NSDictionary <NSString *, NSString *> *)capabilities[FB_CAP_ENVIRNOMENT] ?: @{};
-      [app launch];
+      if (nil != initialUrl) {
+        NSError *openError;
+        didOpenInitialUrl = [XCUIDevice.sharedDevice fb_openUrl:initialUrl
+                                               withApplication:bundleID
+                                                         error:&openError];
+        if (!didOpenInitialUrl) {
+          NSString *errorMsg = [NSString stringWithFormat:@"Cannot open the URL %@ in %@ application. Original error: %@",
+                                initialUrl, bundleID, openError.description];
+          return FBResponseWithStatus([FBCommandStatus sessionNotCreatedError:errorMsg traceback:nil]);
+        }
+      } else {
+        [app launch];
+      }
       if (![app running]) {
         NSString *errorMsg = [NSString stringWithFormat:@"Cannot launch %@ application. Make sure the correct bundle identifier has been provided in capabilities and check the device log for possible crash report occurrences", bundleID];
         return FBResponseWithStatus([FBCommandStatus sessionNotCreatedError:errorMsg
                                                                   traceback:nil]);
       }
     } else if (appState == XCUIApplicationStateRunningBackground && !forceAppLaunch) {
       [app activate];
+      if (nil != initialUrl) {
```

Comment:
Should the app activate be in else clause like launch? Both might not need to be foreground before opening apps via deep link

Reply:
hm, I need to check it one more time. In my tests the app was not restored from the foreground after calling openUrl

Reply:
it worked this time 🤷 . Changed

Reply:
It could depend on the app implementation and iOS internal stuff. perhaps I met similar behavior long before, although don't recall clearly. Then, the app occasionally did not get any notifications by iOS.